Add simple backend

bmwant opened the pull request:
### Rationale

Provide in-memory backend for usage in unittests when no external service is available as per https://github.com/jazzband/django-constance/issues/285. 

It can also be used to cover this https://github.com/jazzband/django-constance/issues/349 use case.

### Usage

Update your test config in the app which uses `django-constance` library with 

```
CONSTANCE_BACKEND = 'constance.backends.simple.SimpleBackend'
```

Run your tests as usually (e.g. `python manage.py test`) using this in-memory backend without external database / redis instance.


### Changes

* Added new `SimpleBackend` in-memory backend
* Moved backend tests to `tests/backends` subdirectory
* Added tests for `mget` backend method

### TODO

* [ ] Update documentation if PR is going to be merged

bmwant commented:
@jezdez @camilonova could you review this?

camilonova commented:
Looks good to me. Please add documentation so we can merge this one 🍺

jezdez commented:
Hey, I haven't looked in detail into the implementation, but I'm thinking we should name this "MemoryBackend" or something similar since "simple" is rarely descriptive enough for users and instead could mislead them about its use case.

Secondly since the main purpose of the backend is to be used for testing, can we make sure that this is thread-safe when accessing the underlying dict?

bmwant commented:
@jezdez I've refactored code to `MemoryBackend` and added lock for thread safety. Python built-in structures and their methods are thread-safe by default, so the only place where race condition can occur is when retrieving multiple values (context switch while other thread modifying a setting), so I've modified `mget` method to reflect that.

jezdez commented on constance/backends/memory.py:
```diff
@@ -0,0 +1,35 @@
+    """
+    def __init__(self):
+        super().__init__()
+        self._storage = {}
```
I'm probably a bit uncertain about threading internals, but I think we should move this into a module global so separate memory backend instances in the individual threads are using the same global storage, since that's the use case of constance.

bmwant replied:
ok, I've got your point, will make corresponding updates

bmwant commented:
@jezdez Django uses locks within `loclmem` because its methods are not atomic (there are a couple of operations over the storage), but I agree we should stay close to that implementation.

I have also tested current solution with
```
def test_same_storage():
    s1 = MemoryBackend()
    s2 = MemoryBackend()
    assert id(s1._storage) == id(s2._storage)

    s1.set('s1-key', 's1-value')
    s2.set('s2-key', 's2-value')

    assert s1.get('s2-key') == s2.get('s2-key')
    assert s2.get('s1-key') == s1.get('s1-key')

    assert s1._storage == s2._storage


def test_same_lock():
    s1 = MemoryBackend()
    s2 = MemoryBackend()
    assert id(s1._lock) == id(s2._lock)
```

and with 

```
from threading import Thread

from constance.backends.memory import MemoryBackend

def target(target_name):
    s = MemoryBackend()
    s.set(target_name, 'value')

    print(s.get(target_name))
    print(id(s._storage))

def test_threads_share_storage():
    s = MemoryBackend()
    t1 = Thread(name='t1', target=target,  args=('t1',))
    t2 = Thread(name='t2', target=target, args=('t2',))

    t1.start()
    t2.start()

    t1.join()
    t2.join()
    # check storage in main thread
    assert len(s._storage) == 2
    assert s.get('t1') == 'value'
    assert s.get('t2') == 'value'
```

to make sure backend works properly and it's equivalent to have `_storage` and `_lock` on class level and on the module level.
